### PR TITLE
Filter assignment metrics by assignment IDs and user h_userid

### PR DESCRIPTION
**Pull request by marcospri**
For:

- https://github.com/hypothesis/lms/issues/6418

### Testing

Play with the filters in something like:

http://localhost:8001/api/dashboard/courses/1826/assignments/metrics?assignment_id=352&assignment_id=353&h_userid=acct:9ea84c4bcfd9329b4495f5b27d8cb6@lms.hypothes.is&h_userid=acct:f6c3ed0edd8a4013bfc8d0c22b9eca@lms.hypothes.is

### CODE REVIEW

File: lms/views/dashboard/api/assignment.py [hunk 1]
```diff
@@ -99,13 +103,16 @@ def course_assignments_metrics(self) -> APIAssignments:
...
-            self.assignment_service.get_assignments(
-                course_id=course.id,
-                instructor_h_userid=current_h_userid,
-                h_userids=filter_by_h_userids,
+        assignments_query = self.assignment_service.get_assignments(
+            course_id=course.id,
+            instructor_h_userid=current_h_userid,
+            h_userids=filter_by_h_userids,
+        )
+        if filter_by_assignment_ids:
```
Comment (marcospri) [inline lms/views/dashboard/api/assignment.py:111]: Arguably this filter could be move into the service itself but in this case I reckon the service is doing the heavy lifting of complicated joins etc and here we just tweak the query slightly to for the view.

Comment (marcospri) [reply lms/views/dashboard/api/assignment.py:111]: Same discussion over: https://github.com/hypothesis/lms/pull/6436#discussion_r1670249819

File: tests/unit/lms/views/dashboard/api/assignment_test.py [hunk 2]
```diff
@@ -140,6 +140,77 @@ def test_course_assignments(
...
+    def test_course_assignments_filtered_by_assignment_ids(
+        self,
+        views,
+        pyramid_request,
+        assignment_service,
+        h_api,
+        db_session,
+        dashboard_service,
+        user_service,
+    ):
```
Comment (robertknight) [inline tests/unit/lms/views/dashboard/api/assignment_test.py:152]: There is a lot of code in this test that isn't directly connected to the change to filter by assignment IDs. It looks like you've got some future refactoring in-flight to reduce this in https://github.com/hypothesis/lms/pull/6437. What's the plan here?

**Review by robertknight — COMMENTED** [review]
Functionally the code change looks quite straightforward. I had a query about the tests.

File: lms/views/dashboard/api/assignment.py [hunk 3]
```diff
@@ -30,6 +30,9 @@ class AssignmentsMetricsSchema(PyramidRequestSchema):
     h_userids = fields.List(fields.Str())
     """Return metrics for these users only."""
 
+    assignment_ids = fields.List(fields.Integer())
+    """Return metrics for these assignments only."""
```
[the remark this answers is not in the document]
Comment (marcospri) [reply lms/views/dashboard/api/assignment.py:34]: Good idea, done the same for h_userid in this PR. 

Will apply the same pattern to all other filters when we merge this.

File: tests/unit/lms/views/dashboard/api/assignment_test.py (hunk 2 above)
Comment (marcospri) [reply tests/unit/lms/views/dashboard/api/assignment_test.py:144]: Merged #6437 into this one so we can look at them combined.

File: tests/unit/lms/services/assignment_test.py [hunk 4]
```diff
@@ -239,55 +239,52 @@ def test_is_member(self, svc, db_session):
         assert svc.is_member(assignment, user.h_userid)
         assert not svc.is_member(assignment, other_user.h_userid)
 
-    def test_get_assignments(self, svc, db_session):
-        assert db_session.scalars(svc.get_assignments()).all()
-
-    def test_get_assignments_by_course(self, svc, db_session, assignment):
+    @pytest.mark.parametrize("instructor_h_userid", [True, False])
```
Comment (marcospri) [inline tests/unit/lms/services/assignment_test.py:242]: This pattern tests all filterings indepently and combined. 

I reckon it makes for a better test. If we merge this I'll apply the same refactor to the user and course services.

File: tests/unit/lms/views/dashboard/api/assignment_test.py [hunk 5]
```diff
@@ -82,18 +83,7 @@ def test_course_assignments(
         user_service.get_users.return_value = (
             select(User).where(User.id.in_([u.id for u in users])).order_by(User.id)
         )
-
-        stats = [
```
Comment (marcospri) [inline tests/unit/lms/views/dashboard/api/assignment_test.py:86]: Moved to common fixture.

File: tests/unit/lms/views/dashboard/api/assignment_test.py (hunk 2 above)
Comment (marcospri) [reply tests/unit/lms/views/dashboard/api/assignment_test.py:144]: For this particular test I changed the assertions to be the most relevant to the test case.

We could move more lines here to fixtures but there's a balance between verbosity in the test vs tightly coupled fixtures.

**Comment by marcospri** [comment]
@robertknight Merged the test refactor PR into this one. That deals with the testing of the service layer filters.

Made some small improvements to the API level test to avoid some verbosity.

File: lms/views/dashboard/api/assignment.py [hunk 6]
```diff
@@ -27,9 +27,12 @@ class AssignmentsMetricsSchema(PyramidRequestSchema):
 
     location = "querystring"
 
-    h_userids = fields.List(fields.Str())
+    h_userids = fields.List(fields.Str(), data_key="h_userid")
```
Comment (marcospri) [inline lms/views/dashboard/api/assignment.py:30]: I reckon the field should still be called with the plural in the schema and python code but using the singular form for each value in the query string.

File: tests/unit/lms/services/assignment_test.py [hunk 7]
```diff
@@ -239,55 +239,52 @@ def test_is_member(self, svc, db_session):
...
+    @pytest.mark.parametrize("course_id", [True, False])
+    @pytest.mark.parametrize("h_userids", [True, False])
+    def test_get_assignments(
```
Comment (robertknight) [inline tests/unit/lms/services/assignment_test.py:245]: This scales as `2^N` where `N` is the number of independent filters. This is OK as long as `N` is small!

File: lms/views/dashboard/api/assignment.py (hunk 6 above)
Comment (robertknight) [reply lms/views/dashboard/api/assignment.py:30]: Yes, that makes sense to me.

File: tests/unit/lms/views/dashboard/api/assignment_test.py [hunk 8]
```diff
@@ -140,6 +130,68 @@ def test_course_assignments(
...
+        db_session.flush()
+
+        assignment_service.get_assignments.return_value = select(Assignment).where(
+            Assignment.id.in_([assignment.id, assignment_with_no_annos.id])
+        )
+        user_service.get_users.return_value = (
+            select(User).where(User.id.in_([u.id for u in users])).order_by(User.id)
+        )
+
+        h_api.get_annotation_counts.return_value = assignments_metrics_reponse
```
Comment (robertknight) [inline tests/unit/lms/views/dashboard/api/assignment_test.py:160]:
```suggestion
        h_api.get_annotation_counts.return_value = assignments_metrics_response
```

File: tests/unit/lms/views/dashboard/api/assignment_test.py [hunk 9]
```diff
@@ -140,6 +130,68 @@ def test_course_assignments(
...
+        self,
+        views,
+        pyramid_request,
+        assignment_service,
+        h_api,
+        db_session,
+        dashboard_service,
+        user_service,
+        assignment,
+        assignments_metrics_reponse,
```
Comment (robertknight) [inline tests/unit/lms/views/dashboard/api/assignment_test.py:143]:
```suggestion
        assignments_metrics_response,
```

File: tests/unit/lms/views/dashboard/api/assignment_test.py [hunk 10]
```diff
@@ -140,6 +130,68 @@ def test_course_assignments(
...
+                        "annotations": sentinel.annotations,
+                        "replies": sentinel.replies,
+                        "last_activity": sentinel.last_activity,
+                    },
+                },
+            ]
+        }
+
+    @pytest.fixture
+    def assignments_metrics_reponse(self, assignment):
```
Comment (robertknight) [inline tests/unit/lms/views/dashboard/api/assignment_test.py:184]:
```suggestion
    def assignments_metrics_response(self, assignment):
```

**Review by robertknight — APPROVED** [review]

**Pull request merged**